packages/lit-html/src/static.ts

```diff
@@ -0,0 +1,77 @@
+import {html as coreHtml, svg as coreSvg, TemplateResult} from './lit-html.js';
+
+/**
+ * Wraps a string so that it behaves like part of the static template
+ * strings instead of a dynamic value.
+ *
+ * This is a very unsafe operation and may break templates if changes
+ * the structure of a template. Do not pass user input to this function
+ * without sanitizing it.
+ */
+export const unsafeStatic = (value: string) => ({
+  _$litStatic$: value,
+});
+
+type StaticValue = ReturnType<typeof unsafeStatic>;
+
+const stringsCache = new WeakMap<TemplateStringsArray, TemplateStringsArray>();
+
+/**
+ * Generates a template literal tag function that returns a TemplateResult with
+ * the given result type.
+ */
+const tag = (coreTag: typeof coreHtml) => (
+  strings: TemplateStringsArray,
+  ...values: unknown[]
+): TemplateResult => {
+  let processedStrings = stringsCache.get(strings);
+  if (processedStrings === undefined) {
+    const newStrings: Array<string> = [];
+    const l = strings.length - 1;
+    let i = 0;
+
+    while (i < l) {
+      let s = strings[i];
+
+      // Collect any unsafeStatic values, and their following template strings
+      // so that we treat a run of template strings and unsafe static values as
+      // a single template string.
+      while (i < l && (values[i] as StaticValue)?._$litStatic$ !== undefined) {
+        s += (values[i] as StaticValue)?._$litStatic$ + strings[++i];
+      }
+      newStrings.push(s);
+      // If by consuming unsafe static values we use the last template string,
+      // then this template has no dynamic bindings.
+      if (i++ > l) {
+        break;
+      }
+    }
+    if (i <= l) {
+      newStrings.push(strings[l]);
```

Comment:
What case is this? `strings[l]` is outside the array range isn't it?

Reply:
This is for any case where the last value isn't static and so the last string wasn't consumed. `strings[l]` is the last element of the array because `l = strings.length - 1`. This works out because the strings array is one longer than the values array. I updated the condition to use `===` and added a comment.